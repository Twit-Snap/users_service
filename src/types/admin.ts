export interface Admin {
    username: string;
    email: string;
}

export interface AdminInfoDto {
    username: string;
    email: string;
    password: string;
}

export interface LoginAdminDto {
    email: string;
    password: string;
}