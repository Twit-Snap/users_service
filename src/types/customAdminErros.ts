export class AlreadyExistError extends Error {
    entityName: string;
    entityEmail: string;
    detail: string

    constructor(entityName: string, entityEmail: string, detail: string) {

        super(`${entityName ? entityName : ''} ${entityEmail ? entityEmail : ''} already exists`);
        this.entityEmail = entityEmail;
        this.entityName = entityName;
        this.detail = detail
        this.name = 'InvalidCredentials';
    }

}

export class ValidationError extends Error {
    field: string;
    detail: string;

    constructor(field: string, detail: string) {
        super(`Validation error: ${field}`);
        this.field = field;
        this.detail = detail;
        this.name = 'ValidationError';
    }
}

export class InvalidCredentialsError extends Error {
    detail: string;

    constructor(detail: string) {

        super(`Email or password are incorrect`);
        this.detail = detail
        this.name = 'InvalidCredentials';
    }

}

export class NotFoundError extends Error {
    entityId: string;
    constructor(entityId: string) {
        super(`${entityId} not found`);
        this.entityId = entityId;
        this.name = 'NotFoundError';
    }
}