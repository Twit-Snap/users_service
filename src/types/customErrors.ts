export class NotFoundError extends Error {
  entityName: string;
  entityId: string;

  constructor(entityName: string, entityId: string) {
    super(`${entityName} not found`);
    this.entityName = entityName;
    this.entityId = entityId;
    this.name = 'NotFoundError';
  }
}

export class ValidationError extends Error {
  field: string;
  detail: string;

  constructor(field: string, detail: string) {
    super(`Validation error: ${field}`);
    this.field = field;
    this.detail = detail;
    this.name = 'ValidationError';
  }
}
