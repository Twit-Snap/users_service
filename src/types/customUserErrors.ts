export class NotFoundError extends Error {
  entityName: string;
  entityId: string;

  constructor(entityName: string, entityId: string) {
    super(`${entityName} not found`);
    this.entityName = entityName;
    this.entityId = entityId;
    this.name = 'NotFoundError';
  }
}

export class ValidationError extends Error {
  field: string;
  detail: string;
  type: string;

  constructor(field: string, detail: string, type: string) {
    super(`Validation error: ${field}`);
    this.field = field;
    this.detail = detail;
    this.name = 'ValidationError';
    this.type = type;
  }
}

export class EntityAlreadyExistsError extends Error {
  entityName: string;

  constructor(entityName: string) {
    super(`${entityName} already exists`);
    this.entityName = entityName;
    this.name = 'EntityAlreadyExistsError';
  }
}

export class AuthenticationError extends Error {

  constructor() {
    super('Authentication error');
    this.name = 'AuthenticationError';
  }
}