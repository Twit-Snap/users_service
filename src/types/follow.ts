import { TwitUser } from 'user';

export type FollowResponse = {
  username: string;
  followedUsername: string;
  createdAt: string;
};

export type FollowReturn = {
  createdAt: string;
};

export type FollowersResponse = {
  follower: Omit<TwitUser, 'id'>;
  createdAt: string;
};

export type FollowersReturn = {
  id: number;
  username: string;
  name: string;
  createdAt: string;
};
