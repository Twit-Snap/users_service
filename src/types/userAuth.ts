import { User } from 'user';

export type UserWithToken = User & { token: string };

export interface UserRegisterDto {
  username: string;
  email: string;
  name: string;
  lastname: string;
  birthdate: Date;
  password: string;
  registrationTime: number;
}

export interface UserSSORegisterDto {
  token: string;
  uid: string;
  providerId: string;
  username: string;
  birthdate: Date;
}

export interface UserSSOLoginDto {
  token: string;
  uid: string;
}

export type UserRegisterRepository = Omit<UserRegisterDto, 'password'|'registrationTime' > & {
  ssoUid?: string | null;
  ssoProviderId?: string | null;
  password?: string | null;
  profilePicture?: string | null;
};
export interface IUserAuthService {
  login(emailOrUsername: string, password: string): Promise<UserWithToken>;
  register(user: UserRegisterDto): Promise<UserWithToken>;
}
