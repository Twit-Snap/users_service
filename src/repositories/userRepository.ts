import { EntityAlreadyExistsError } from 'customErrors';
import { Pool, QueryResult } from 'pg';
import { IUserRepository, UserRegisterDto, User, UserWithPassword } from 'user';

export class UserRepository implements IUserRepository{
  private pool: Pool;

  constructor(pool: Pool) {
    this.pool = pool;
  }
  async findByEmailOrUsername(emailOrUsername: string): Promise<UserWithPassword | null> {
    const query = 'SELECT id, username, email, name, lastname, birthdate, password, created_at AS createdAt FROM users WHERE email = $1 OR username = $1';
    const result = await this.pool.query<UserWithPassword>(query, [emailOrUsername]);
    if (result.rows.length === 0) {
      return null;
    }
    return result.rows[0];
  }

  async getList(): Promise<User[] | null> {
    const query = 'SELECT id, username, email, name, lastname, birthdate, created_at AS createdAt FROM users';
    const result = await this.pool.query<User>(query);
    if (result.rows.length === 0) {
      return null;
    }
    return result.rows;
  }

  async get(id: number): Promise<User | null> {
    const query = 'SELECT id, username, email, name, lastname, birthdate, created_at AS createdAt FROM users WHERE id = $1';
    const result = await this.pool.query<User>(query, [id]);
    if (result.rows.length === 0) {
      return null;
    }
    return result.rows[0];
  }

  async create(userData: UserRegisterDto): Promise<User> {
    const { username, email, name, lastname, birthdate, password } = userData;
    const query = `
      INSERT INTO users (username, email, name, lastname, birthdate, password)
      VALUES ($1, $2, $3, $4, $5, $6)
      RETURNING id, username, email, name, lastname, birthdate, created_at AS "createdAt"
    `;
    
    try {
      const result = await this.pool.query<User>(query, [
        username,
        email,
        name,
        lastname,
        birthdate,
        password
      ]);
      return result.rows[0];
    } catch (error) {
      console.error(error);
      const errorAux = error as { code: string, constraint: string };
      if (errorAux.code === '23505') { // PostgreSQL unique constraint violation error code
        if (errorAux.constraint?.includes('username')) {
          throw new EntityAlreadyExistsError('Username');
        } else if (errorAux.constraint?.includes('email')) {
          throw new EntityAlreadyExistsError('Email');
        }
      }
      // If it's not a unique constraint violation, re-throw the original error
      throw error;
    }
  }

}
