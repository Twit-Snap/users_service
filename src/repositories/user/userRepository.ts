import { FollowersResponse, FollowReturn, GetAllFollowsParams } from 'follow';
import { Pool } from 'pg';
import {
  GetUserParams,
  GetUsersListParams,
  Interest,
  IUserRepository,
  ModifiableUser,
  OnlyExpoToken,
  User,
  UserWithPassword
} from 'user';
import { UserRegisterRepository } from 'userAuth';
import { EntityAlreadyExistsError } from '../../types/customErrors';
import { userCamelColumnsToSnake } from '../../utils/user';
import { DatabasePool } from '../db';

export class UserRepository implements IUserRepository {
  private pool: Pool;
  private readonly selectUserFields = `id, username, email, name, lastname, birthdate, users.created_at AS "createdAt", sso_uid as "ssoUid", provider_id as "providerId", profile_picture as "profilePicture", is_private AS "isPrivate", is_blocked AS "isBlocked", expo_token AS "expoToken", phone_number AS "phoneNumber", verified, background_picture AS "backgroundPicture"`;
  private readonly reducedUserFields =
    'id, username, name, profile_picture AS "profilePicture", is_private AS "isPrivate", expo_token AS "expoToken", background_picture AS "backgroundPicture"';

  constructor(pool?: Pool) {
    this.pool = pool || DatabasePool.getInstance();
  }
  async findByEmailOrUsername(emailOrUsername: string): Promise<UserWithPassword | null> {
    const query = `SELECT ${this.selectUserFields}, password FROM users WHERE email = $1 OR username = $1`;
    const result = await this.pool.query<UserWithPassword>(query, [emailOrUsername]);
    if (result.rows.length === 0) {
      return null;
    }
    return result.rows[0];
  }
  async findBySSOuid(uid: string): Promise<User | null> {
    const query = `SELECT ${this.selectUserFields} FROM users WHERE sso_uid = $1`;
    const result = await this.pool.query<User>(query, [uid]);
    if (result.rows.length === 0) {
      return null;
    }
    return result.rows[0];
  }

  async getAmount(params: GetUsersListParams): Promise<number> {
    var offset = '';
    var queryParams: (number | string)[] = [];

    if (params.createdAt) {
      queryParams.push(params.createdAt);
      offset = `WHERE created_at ${params.equalDate ? '<=' : '<'} $${queryParams.length}::timestamptz ${params.equalDate ? ` + interval '0.999999 seconds'` : ''}`;
    }

    const query = `SELECT COUNT(*) FROM users ${offset}`;
    const result = await this.pool.query(query, queryParams);
    return result.rows[0];
  }

  async getList(params: GetUsersListParams): Promise<User[]> {
    var queryParams: (string | number)[] = [`%${params.has}%`];
    var offset = '';
    var limit = '';

    if (params.createdAt) {
      queryParams.push(params.createdAt);
      offset = ` AND created_at ${params.equalDate ? '<=' : '<'} $${queryParams.length}::timestamptz ${params.equalDate ? ` + interval '0.999999 seconds'` : ''}`;
    }

    if (params.limit) {
      queryParams.push(params.limit);
      queryParams.push(params.offset);
      limit = ` LIMIT $${queryParams.length - 1} OFFSET $${queryParams.length}`;
    }

    const query = `
    SELECT ${this.selectUserFields}
    FROM users
    WHERE (username ILIKE $1 OR name ILIKE $1)${offset}
    ORDER BY created_at DESC
    ${limit}`;

    const result = await this.pool.query<User>(query, queryParams);
    return result.rows;
  }

  async get(id: number): Promise<User | null> {
    const query = `SELECT ${this.selectUserFields} FROM users WHERE id = $1`;
    const result = await this.pool.query<User>(query, [id]);
    if (result.rows.length === 0) {
      return null;
    }
    return result.rows[0];
  }

  async create(userData: UserRegisterRepository): Promise<User> {
    const {
      username,
      email,
      name,
      lastname,
      password,
      profilePicture,
      ssoProviderId,
      ssoUid,
      expoToken,
      phoneNumber
    } = userData;
    const birthdate = new Date(userData.birthdate).toISOString();

    const query = `
      INSERT INTO users (username, email, name, lastname, birthdate, password, profile_picture, sso_uid, provider_id, expo_token, phone_number)
      VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
      RETURNING ${this.selectUserFields}
    `;

    try {
      const result = await this.pool.query<User>(query, [
        username,
        email,
        name,
        lastname,
        birthdate,
        password,
        profilePicture,
        ssoUid,
        ssoProviderId,
        expoToken,
        phoneNumber
      ]);
      return result.rows[0];
    } catch (error) {
      console.error(error);
      const errorAux = error as { code: string; constraint: string };
      if (errorAux.code === '23505') {
        // PostgreSQL unique constraint violation error code
        if (errorAux.constraint?.includes('username')) {
          throw new EntityAlreadyExistsError('Username', 'Username is already in use');
        } else if (errorAux.constraint?.includes('email')) {
          throw new EntityAlreadyExistsError('Email', 'Email is already in use');
        } else if (errorAux.constraint?.includes('sso_uid')) {
          throw new EntityAlreadyExistsError('SSOUid', 'SSO UID is already in use');
        } else if (errorAux.constraint?.includes('phone_number')) {
          throw new EntityAlreadyExistsError('phoneNumber', 'Phone number is already in use');
        }
      }
      // If it's not a unique constraint violation, re-throw the original error
      throw error;
    }
  }

  async getByUsername(username: string, params?: GetUserParams) {
    const query = `SELECT ${params?.reduce ? this.reducedUserFields : this.selectUserFields} FROM users WHERE username = $1`;
    const result = await this.pool.query<User>(query, [username]);
    if (result.rows.length === 0) {
      return null;
    }
    return result.rows[0];
  }

  async createFollow(userId: number, followId: number): Promise<FollowReturn> {
    const query = `
      INSERT INTO follows (userId, followedId)
      VALUES ($1, $2)
      RETURNING userId, followedId, created_at AS "createdAt"
    `;
    try {
      const result = await this.pool.query<FollowReturn>(query, [userId, followId]);
      return result.rows[0];
    } catch (error) {
      console.error(error);

      const errorAux = error as { code: string; constraint: string };
      if (errorAux.code === '23505') {
        // PostgreSQL unique constraint violation error code
        throw new EntityAlreadyExistsError('Follow', `(user, followed) key alredy exists`);
      }

      throw error;
    }
  }

  async deleteFollow(userId: number, followId: number): Promise<void> {
    const query = `
      DELETE FROM follows
      WHERE userId = $1 AND followedId = $2
    `;

    try {
      await this.pool.query<FollowReturn>(query, [userId, followId]);
    } catch (error) {
      console.error(error);
      throw error;
    }
  }

  async getFollow(userId: number, followId: number): Promise<FollowReturn | undefined> {
    const query = `
      SELECT userId, followedId, created_at AS "createdAt"
      FROM follows
      WHERE userId = $1 AND followedId = $2
    `;

    try {
      const result = await this.pool.query<FollowReturn>(query, [userId, followId]);
      return result.rows[0];
    } catch (error) {
      console.error(error);
      throw error;
    }
  }

  async getFollows(userId: number, params: GetAllFollowsParams): Promise<FollowersResponse[]> {
    var queryParams: (number | string)[] = [userId, `%${params.has}%`];

    const condition = params.byFollowers ? 'followedId = $1' : 'userId = $1';
    const join = params.byFollowers ? 'userId' : 'followedId';
    var offset = '';
    var limit = '';

    if (params.createdAt) {
      queryParams.push(params.createdAt);
      offset = ` AND follows.created_at < $${queryParams.length}`;
    }

    if (params.limit) {
      queryParams.push(Number(params.limit));
      limit = ` LIMIT $${queryParams.length}`;
    }

    const query = `
      SELECT id, username, name, follows.created_at AS "followCreatedAt"
      FROM follows
      INNER JOIN users ON follows.${join} = users.id
      WHERE ${condition}${offset} AND (username ILIKE $2 OR name ILIKE $2)
      ORDER BY follows.created_at DESC
      ${limit}
    `;

    try {
      const result = await this.pool.query<FollowersResponse>(query, queryParams);
      return result.rows;
    } catch (error) {
      console.error(error);
      throw error;
    }
  }

  async getFollowsFullData(userId: number, params: GetAllFollowsParams): Promise<User[]> {
    var queryParams: (number | string)[] = [userId, `%${params.has}%`];

    const condition = params.byFollowers ? 'followedId = $1' : 'userId = $1';
    const join = params.byFollowers ? 'userId' : 'followedId';
    var offset = '';
    var limit = '';

    if (params.createdAt) {
      queryParams.push(params.createdAt);
      offset = ` AND follows.created_at < $${queryParams.length}`;
    }

    if (params.limit) {
      queryParams.push(Number(params.limit));
      limit = ` LIMIT $${queryParams.length}`;
    }

    const query = `
      SELECT ${this.selectUserFields}
      FROM follows
      INNER JOIN users ON follows.${join} = users.id
      WHERE ${condition}${offset} AND (username ILIKE $2 OR name ILIKE $2)
      ORDER BY follows.created_at DESC
      ${limit}
    `;

    try {
      const result = await this.pool.query<User>(query, queryParams);
      return result.rows;
    } catch (error) {
      console.error(error);
      throw error;
    }
  }

  async modifyUser(userId: number, newValues: ModifiableUser): Promise<User> {
    let queryParams: (string | number | boolean)[] = [userId];

    const setVars: string = Object.entries(newValues)
      .map(([key, val]) => {
        queryParams.push(val);
        const snakeKey = userCamelColumnsToSnake(key);
        return `${snakeKey} = $${queryParams.length}`;
      })
      .join(', ');

    const query = `
      UPDATE users
      SET ${setVars}
      WHERE id = $1
      RETURNING ${this.selectUserFields}
    `;

    try {
      const result = await this.pool.query<User>(query, queryParams);
      return result.rows[0];
    } catch (error) {
      console.error(error);
      const errorAux = error as { code: string; constraint: string };
      if (errorAux.code === '23505') {
        // PostgreSQL unique constraint violation error code
        if (errorAux.constraint?.includes('username')) {
          throw new EntityAlreadyExistsError('Username', 'Username is already in use');
        } else if (errorAux.constraint?.includes('email')) {
          throw new EntityAlreadyExistsError('Email', 'Email is already in use');
        }
      }
      // If it's not a unique constraint violation, re-throw the original error
      throw error;
    }
  }

  async putExpoToken(userId: number, expoToken: string): Promise<void> {
    const query = `
      UPDATE users
      SET expo_token = $2
      WHERE id = $1
    `;

    try {
      await this.pool.query(query, [userId, expoToken]);
    } catch (error) {
      console.error(error);
      // If it's not a unique constraint violation, re-throw the original error
      throw error;
    }
  }

  async updateLocation(
    username: string,
    location: { latitude: number; longitude: number }
  ): Promise<void> {
    const query = `
      UPDATE users
      SET latitude = $1, longitude = $2
      WHERE username = $3
    `;
    await this.pool.query(query, [location.latitude, location.longitude, username]);
  }

  async getAllExpoTokens(senderId: number): Promise<OnlyExpoToken[]> {
    const query = `
      SELECT expo_token AS "expoToken"
      FROM users
      WHERE id <> $1 and expo_token IS NOT NULL
      `;

    const result = await this.pool.query(query, [senderId]);
    return result.rows;
  }

  async getAllInterests(): Promise<Interest[]> {
    const query = 'SELECT id, name, parent_id AS "parentId", emoji FROM interests';
    const result = await this.pool.query(query);
    return result.rows;
  }

  async getUserInterests(userId: number): Promise<Interest[]> {
    const query = `
      SELECT i.id, i.name, i.parent_id AS "parentId", i.emoji
      FROM user_interests ui
      JOIN interests i ON ui.interest_id = i.id
      WHERE ui.user_id = $1
    `;
    const result = await this.pool.query<Interest>(query, [userId]);
    return result.rows;
  }

  async associateInterestsToUser(userId: number, interests: number[]): Promise<boolean> {
    // in the service we check that interests is not empty and are numbers, so this is safe
    const valuesQuery = interests.map((interestId) => `(${userId}, ${interestId})`).join(', ');
    const query = `
      INSERT INTO user_interests (user_id, interest_id)
      VALUES ${valuesQuery}
      ON CONFLICT (user_id, interest_id) DO NOTHING
    `;
    const result = await this.pool.query(query);
    return result.rowCount !== null && result.rowCount > 0;
  }

  async updatePassword(userId: number, password: string): Promise<void> {
    const query = `UPDATE users SET password = $1 WHERE id = $2`;
    await this.pool.query(query, [password, userId]);
  }
}
