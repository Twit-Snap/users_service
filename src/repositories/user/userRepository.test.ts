/* eslint-disable @typescript-eslint/no-explicit-any */
import { Pool } from 'pg';
import { UserRegisterDto } from 'userAuth';
import { EntityAlreadyExistsError } from '../../types/customErrors';
import { UserRepository } from './userRepository';

function interpolateQuery(query: string, params: any[]): string {
  return query.replace(/\$\d+/g, (match) => {
    const index = parseInt(match.substr(1)) - 1;
    const value = params[index];
    return typeof value === 'string' ? `'${value}'` : value;
  });
}

describe('UserRepository', () => {
  let userRepository: UserRepository;
  let mockPool: {
    query: jest.Mock;
  };

  beforeEach(() => {
    mockPool = {
      query: jest.fn()
    };
    userRepository = new UserRepository(mockPool as unknown as Pool);
  });

  describe('findByEmailOrUsername', () => {
    it('should find user by email or username', async () => {
      const mockUser = { id: 1, username: 'testuser', email: 'test@example.com' };
      mockPool.query.mockResolvedValueOnce({ rows: [mockUser] });

      const result = await userRepository.findByEmailOrUsername('testuser');
      expect(result).toMatchSnapshot('User result');
      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('SQL query');
      const [query, params] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, params);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });

    it('should return null when user not found', async () => {
      mockPool.query.mockResolvedValueOnce({ rows: [] });

      const result = await userRepository.findByEmailOrUsername('nonexistent');
      expect(result).toMatchSnapshot('User result');
      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('SQL query');
      const [query, params] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, params);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });
  });

  describe('getList', () => {
    it('should get list of users', async () => {
      const mockUsers = [
        { id: 1, username: 'user1', email: 'user1@example.com' },
        { id: 2, username: 'user2', email: 'user2@example.com' }
      ];
      mockPool.query.mockResolvedValueOnce({ rows: mockUsers });

      const result = await userRepository.getList();
      expect(result).toMatchSnapshot('Users list result');
      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('SQL query');
      const [query, params] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, params);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });

    it('should return null when no users found', async () => {
      mockPool.query.mockResolvedValueOnce({ rows: [] });

      const result = await userRepository.getList();
      expect(result).toMatchSnapshot('Users list result');
      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('SQL query');
      const [query, params] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, params);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });
  });

  describe('get', () => {
    it('should get user by id', async () => {
      const mockUser = { id: 1, username: 'testuser', email: 'test@example.com' };
      mockPool.query.mockResolvedValueOnce({ rows: [mockUser] });

      const result = await userRepository.get(1);
      expect(result).toMatchSnapshot('User result');
      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('SQL query');
      const [query, params] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, params);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });

    it('should return null when user not found', async () => {
      mockPool.query.mockResolvedValueOnce({ rows: [] });

      const result = await userRepository.get(999);
      expect(result).toMatchSnapshot('User result');
      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('SQL query');
      const [query, params] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, params);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });
  });

  describe('create', () => {
    it('should create a new user', async () => {
      const mockUser = { id: 1, username: 'newuser', email: 'new@example.com' };
      mockPool.query.mockResolvedValueOnce({ rows: [mockUser] });

      const userData: UserRegisterDto = {
        username: 'newuser',
        email: 'new@example.com',
        name: 'New',
        lastname: 'User',
        birthdate: new Date('2000-01-01'),
        password: 'password123'
      };

      const result = await userRepository.create(userData);
      expect(result).toMatchSnapshot('Created user result');
      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('SQL query');
      const [query, params] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, params);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });

    it('should throw EntityAlreadyExistsError for duplicate username', async () => {
      const error = new Error('duplicate key value violates unique constraint');
      (error as any).code = '23505';
      (error as any).constraint = 'username_unique';
      mockPool.query.mockRejectedValueOnce(error);

      const userData: UserRegisterDto = {
        username: 'existinguser',
        email: 'new@example.com',
        name: 'New',
        lastname: 'User',
        birthdate: new Date('2000-01-01'),
        password: 'password123'
      };

      await expect(userRepository.create(userData)).rejects.toThrowError(EntityAlreadyExistsError);
      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('SQL query');
      const [query, params] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, params);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });

    it('should throw EntityAlreadyExistsError for duplicate email', async () => {
      const error = new Error('duplicate key value violates unique constraint');
      (error as any).code = '23505';
      (error as any).constraint = 'email_unique';
      mockPool.query.mockRejectedValueOnce(error);

      const userData: UserRegisterDto = {
        username: 'newuser',
        email: 'existing@example.com',
        name: 'New',
        lastname: 'User',
        birthdate: new Date('2000-01-01'),
        password: 'password123'
      };

      await expect(userRepository.create(userData)).rejects.toThrowError(EntityAlreadyExistsError);
      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('SQL query');
      const [query, params] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, params);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });
  });

  describe('getByUsername', () => {
    it('should get user by username', async () => {
      const mockUser = { username: 'testuser', email: 'test@example.com' };
      mockPool.query.mockResolvedValueOnce({ rows: [mockUser] });

      const result = await userRepository.getByUsername('testuser');
      expect(result).toMatchSnapshot('User result');
      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('SQL query');
      const [query, params] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, params);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });

    it('should return null when user not found', async () => {
      mockPool.query.mockResolvedValueOnce({ rows: [] });

      const result = await userRepository.getByUsername('nonexistent');
      expect(result).toMatchSnapshot('User result');
      expect(mockPool.query.mock.calls[0]).toMatchSnapshot('SQL query');
      const [query, params] = mockPool.query.mock.calls[0];
      const interpolatedQuery = interpolateQuery(query, params);
      expect(interpolatedQuery).toMatchSnapshot('Interpolated SQL query');
    });
  });
});
