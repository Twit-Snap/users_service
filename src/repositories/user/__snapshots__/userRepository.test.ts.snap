// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRepository create should create a new user: Created user result 1`] = `
{
  "email": "new@example.com",
  "id": 1,
  "username": "newuser",
}
`;

exports[`UserRepository create should create a new user: Interpolated SQL query 1`] = `
"
      INSERT INTO users (username, email, name, lastname, birthdate, password)
      VALUES ('newuser', 'new@example.com', 'New', 'User', '2000-01-01T00:00:00.000Z', 'password123')
      RETURNING id, username, email, name, lastname, birthdate, created_at AS "createdAt"
    "
`;

exports[`UserRepository create should create a new user: SQL query 1`] = `
[
  "
      INSERT INTO users (username, email, name, lastname, birthdate, password)
      VALUES ($1, $2, $3, $4, $5, $6)
      RETURNING id, username, email, name, lastname, birthdate, created_at AS "createdAt"
    ",
  [
    "newuser",
    "new@example.com",
    "New",
    "User",
    "2000-01-01T00:00:00.000Z",
    "password123",
  ],
]
`;

exports[`UserRepository create should throw EntityAlreadyExistsError for duplicate email: Interpolated SQL query 1`] = `
"
      INSERT INTO users (username, email, name, lastname, birthdate, password)
      VALUES ('newuser', 'existing@example.com', 'New', 'User', '2000-01-01T00:00:00.000Z', 'password123')
      RETURNING id, username, email, name, lastname, birthdate, created_at AS "createdAt"
    "
`;

exports[`UserRepository create should throw EntityAlreadyExistsError for duplicate email: SQL query 1`] = `
[
  "
      INSERT INTO users (username, email, name, lastname, birthdate, password)
      VALUES ($1, $2, $3, $4, $5, $6)
      RETURNING id, username, email, name, lastname, birthdate, created_at AS "createdAt"
    ",
  [
    "newuser",
    "existing@example.com",
    "New",
    "User",
    "2000-01-01T00:00:00.000Z",
    "password123",
  ],
]
`;

exports[`UserRepository create should throw EntityAlreadyExistsError for duplicate username: Interpolated SQL query 1`] = `
"
      INSERT INTO users (username, email, name, lastname, birthdate, password)
      VALUES ('existinguser', 'new@example.com', 'New', 'User', '2000-01-01T00:00:00.000Z', 'password123')
      RETURNING id, username, email, name, lastname, birthdate, created_at AS "createdAt"
    "
`;

exports[`UserRepository create should throw EntityAlreadyExistsError for duplicate username: SQL query 1`] = `
[
  "
      INSERT INTO users (username, email, name, lastname, birthdate, password)
      VALUES ($1, $2, $3, $4, $5, $6)
      RETURNING id, username, email, name, lastname, birthdate, created_at AS "createdAt"
    ",
  [
    "existinguser",
    "new@example.com",
    "New",
    "User",
    "2000-01-01T00:00:00.000Z",
    "password123",
  ],
]
`;

exports[`UserRepository findByEmailOrUsername should find user by email or username: Interpolated SQL query 1`] = `"SELECT id, username, email, name, lastname, birthdate, password, created_at AS "createdAt" FROM users WHERE email = 'testuser' OR username = 'testuser'"`;

exports[`UserRepository findByEmailOrUsername should find user by email or username: SQL query 1`] = `
[
  "SELECT id, username, email, name, lastname, birthdate, password, created_at AS "createdAt" FROM users WHERE email = $1 OR username = $1",
  [
    "testuser",
  ],
]
`;

exports[`UserRepository findByEmailOrUsername should find user by email or username: User result 1`] = `
{
  "email": "test@example.com",
  "id": 1,
  "username": "testuser",
}
`;

exports[`UserRepository findByEmailOrUsername should return null when user not found: Interpolated SQL query 1`] = `"SELECT id, username, email, name, lastname, birthdate, password, created_at AS "createdAt" FROM users WHERE email = 'nonexistent' OR username = 'nonexistent'"`;

exports[`UserRepository findByEmailOrUsername should return null when user not found: SQL query 1`] = `
[
  "SELECT id, username, email, name, lastname, birthdate, password, created_at AS "createdAt" FROM users WHERE email = $1 OR username = $1",
  [
    "nonexistent",
  ],
]
`;

exports[`UserRepository findByEmailOrUsername should return null when user not found: User result 1`] = `null`;

exports[`UserRepository get should get user by id: Interpolated SQL query 1`] = `"SELECT id, username, email, name, lastname, birthdate, created_at AS "createdAt" FROM users WHERE id = 1"`;

exports[`UserRepository get should get user by id: SQL query 1`] = `
[
  "SELECT id, username, email, name, lastname, birthdate, created_at AS "createdAt" FROM users WHERE id = $1",
  [
    1,
  ],
]
`;

exports[`UserRepository get should get user by id: User result 1`] = `
{
  "email": "test@example.com",
  "id": 1,
  "username": "testuser",
}
`;

exports[`UserRepository get should return null when user not found: Interpolated SQL query 1`] = `"SELECT id, username, email, name, lastname, birthdate, created_at AS "createdAt" FROM users WHERE id = 999"`;

exports[`UserRepository get should return null when user not found: SQL query 1`] = `
[
  "SELECT id, username, email, name, lastname, birthdate, created_at AS "createdAt" FROM users WHERE id = $1",
  [
    999,
  ],
]
`;

exports[`UserRepository get should return null when user not found: User result 1`] = `null`;

exports[`UserRepository getByUsername should get user by username: Interpolated SQL query 1`] = `"SELECT username, email, name, lastname, birthdate, created_at AS "createdAt" FROM users WHERE username = 'testuser'"`;

exports[`UserRepository getByUsername should get user by username: SQL query 1`] = `
[
  "SELECT username, email, name, lastname, birthdate, created_at AS "createdAt" FROM users WHERE username = $1",
  [
    "testuser",
  ],
]
`;

exports[`UserRepository getByUsername should get user by username: User result 1`] = `
{
  "email": "test@example.com",
  "username": "testuser",
}
`;

exports[`UserRepository getByUsername should return null when user not found: Interpolated SQL query 1`] = `"SELECT username, email, name, lastname, birthdate, created_at AS "createdAt" FROM users WHERE username = 'nonexistent'"`;

exports[`UserRepository getByUsername should return null when user not found: SQL query 1`] = `
[
  "SELECT username, email, name, lastname, birthdate, created_at AS "createdAt" FROM users WHERE username = $1",
  [
    "nonexistent",
  ],
]
`;

exports[`UserRepository getByUsername should return null when user not found: User result 1`] = `null`;

exports[`UserRepository getList should get list of users: Interpolated SQL query 1`] = `"SELECT id, username, email, name, lastname, birthdate, created_at AS "createdAt" FROM users"`;

exports[`UserRepository getList should get list of users: SQL query 1`] = `
[
  "SELECT id, username, email, name, lastname, birthdate, created_at AS "createdAt" FROM users",
]
`;

exports[`UserRepository getList should get list of users: Users list result 1`] = `
[
  {
    "email": "user1@example.com",
    "id": 1,
    "username": "user1",
  },
  {
    "email": "user2@example.com",
    "id": 2,
    "username": "user2",
  },
]
`;

exports[`UserRepository getList should return null when no users found: Interpolated SQL query 1`] = `"SELECT id, username, email, name, lastname, birthdate, created_at AS "createdAt" FROM users"`;

exports[`UserRepository getList should return null when no users found: SQL query 1`] = `
[
  "SELECT id, username, email, name, lastname, birthdate, created_at AS "createdAt" FROM users",
]
`;

exports[`UserRepository getList should return null when no users found: Users list result 1`] = `null`;
